beetsplug/hook.py

```diff
@@ -0,0 +1,83 @@
+# This file is part of beets.
+# Copyright 2015, Adrian Sampson.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""Allows custom commands to be run when an event is emitted by beets"""
+from __future__ import (division, absolute_import, print_function,
+                        unicode_literals)
+
+import subprocess
+
+from beets.ui import _arg_encoding
+from beets.plugins import BeetsPlugin
+
+
+def create_hook_function(log, event, command, shell, substitute_args):
+    def hook_function(**kwargs):
+        hook_command = command
+
+        for key in substitute_args:
+            if key in kwargs:
+                hook_command = hook_command.replace(substitute_args[key],
+                                                    unicode(kwargs[key],
+                                                            _arg_encoding()))
+
+        log.debug('Running command {0} for event {1}', hook_command, event)
+
+        subprocess.Popen(hook_command, shell=shell).wait()
```

Comment:
We'll need to encode the command and arguments before running them (for Unix platforms).